algosdk/abi/type.py

```diff
@@ -0,0 +1,1026 @@
+        """
+        pass
+
+
+class UintType(Type):
+    """
+    Represents an Uint ABI Type for encoding.
+
+    Args:
+        type_id (BaseType): type of ABI argument, as defined by the BaseType class above.
+        bit_size (int, optional): size of a uint type, e.g. for a uint8, the bit_size is 8.
+
+    Attributes:
+        type_id (BaseType)
+        bit_size (int)
+    """
+
+    def __init__(self, type_size) -> None:
+        if (
+            not isinstance(type_size, int)
+            or type_size % 8 != 0
+            or type_size < 8
+            or type_size > 512
+        ):
+            raise error.ABITypeError(
+                "unsupported uint bitSize: {}".format(type_size)
+            )
+        self.abi_type_id = BaseType.Uint
+        self.bit_size = type_size
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, UintType):
+            return False
+        return (
+            self.abi_type_id == other.abi_type_id
+            and self.bit_size == other.bit_size
+        )
+
+    def __str__(self):
+        return "uint{}".format(str(self.bit_size))
+
+    def byte_len(self):
+        return self.bit_size // 8
+
+    def is_dynamic(self):
+        return False
+
+    def encode(self, value):
+        assert isinstance(value, int)
+        if value >= (2 ** self.bit_size) or value < 0:
+            raise error.ABIEncodingError(
+                "value {} is negative or is too big to fit in size {}".format(
+                    value, self.bit_size
+                )
+            )
+        return (value).to_bytes(self.bit_size // 8, byteorder="big")
+
+    def decode(self, value_string):
+        if (
+            not (
+                isinstance(value_string, bytes)
+                or isinstance(value_string, bytearray)
+            )
+            or len(value_string) != self.bit_size // 8
+        ):
+            raise error.ABIEncodingError(
+                "value string must be in bytes and correspond to a uint{}: {}".format(
+                    self.bit_size, value_string
+                )
+            )
+        # Convert bytes into an unsigned integer
+        return int.from_bytes(value_string, byteorder="big", signed=False)
+
+
+class ByteType(Type):
+    """
+    Represents a Byte ABI Type for encoding.
+    """
+
+    def __init__(self) -> None:
+        self.abi_type_id = BaseType.Byte
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, ByteType):
+            return False
+        return self.abi_type_id == other.abi_type_id
+
+    def __str__(self):
+        return "byte"
+
+    def byte_len(self):
+        return 1
+
+    def is_dynamic(self):
+        return False
+
+    def encode(self, value):
+        """
+        Encode a single byte or a uint8
+        """
+        # Enforce ByteType to only accept bytes or int values
+        if isinstance(value, bytes):
+            value = int.from_bytes(value, byteorder="big")
```

Comment:
I don't think there's a need to support `bytes` as an encodable value for this type, since the type represents just a single byte.

But I do think there's a lot of value in supporting `bytes` as an encodable type for dynamic/static ABI arrays of bytes. I believe that's already possible with your existing code, since it appears `TypleType.encode` can accept `bytes` as an input. Though I think you should add a check there to ensure an input type of `bytes` or `bytearray` is only accepted if all of the tuple child types are `ByteType`.

Reply:
> I thought it might be useful to encode characters like `b"a"`

I personally think this is a little misleading, since `b"a"` is really a 1 element byte array. But python is a bit weird about string characters in general (`"a"[0] == "a"`!), so if you think this is useful, it's fine with me. But I would suggest you add an explicit check that the length is 1, so we fail early if `b"ab"` is passed.

> Also the current String encoding relies on the `ByteType.encode()` to encode [characters byte by byte](https://github.com/algorand/py-algorand-sdk/blob/e4fc085c28357dafda29d05fb7d5b308609bfd83/algosdk/abi/string_type.py#L37).

True but it just as easily could do `value_array.append(val)`

> For the second point, do you mean like: if you have `(bytes,bytes)`, the input can either be like `bytes([1, 2])` or `[b"\x1", b"\x2"]`?

Yes, it would be very convent to be able to directly pass in `b"\x01\x02\x03"` instead of `[1, 2, 3]` (or `[b"\x01", b"\x02", b"\x03"]` if we allow treating length 1 `bytes` as a single byte)